From 79c02fe10c5290550088e5a4655ff9c7a6583cb2 Mon Sep 17 00:00:00 2001
From: Jiffin Tony Thottan <jthottan@redhat.com>
Date: Tue, 25 Aug 2015 17:02:58 +0530
Subject: [PATCH] FSAL_GLUSTER : enable subdir mount for symlinks

Previously in v3 subdir mount will fail if path contains links.
This patch will fix the same

Change-Id: I396e3d72843a7d1530cb0d1f66ba54cbf9c74043
Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>
---
 src/CMakeLists.txt             |    7 +++++++
 src/FSAL/FSAL_GLUSTER/export.c |    8 +++++++-
 src/FSAL/FSAL_GLUSTER/handle.c |   10 +++++++---
 src/include/config-h.in.cmake  |    1 +
 4 files changed, 22 insertions(+), 4 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 4e7d443..c50eec9 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -513,6 +513,13 @@ if(USE_FSAL_GLUSTER)
 	message(STATUS "Could not find libacl, disabling GLUSTER fsal build")
     endif(HAVE_ACL_H)
   endif(USE_FSAL_GLUSTER)
+  pkg_check_modules(GLFS_H_LOOKUPAT_WITH_FOLLOW glusterfs-api>=7.3.7.1)
+  if(NOT GLFS_H_LOOKUPAT_WITH_FOLLOW_FOUND)
+	set(USE_GLUSTER_SYMLINK_MOUNT OFF)
+	message(WARNING "Disabling symlink mount for FSAL GLUSTER")
+  else()
+	set(USE_GLUSTER_SYMLINK_MOUNT ON)
+  endif(NOT GLFS_H_LOOKUPAT_WITH_FOLLOW_FOUND)
 endif(USE_FSAL_GLUSTER)
 
 if(USE_FSAL_CEPH)
diff --git a/src/FSAL/FSAL_GLUSTER/export.c b/src/FSAL/FSAL_GLUSTER/export.c
index 598b00a..ea560bc 100644
--- a/src/FSAL/FSAL_GLUSTER/export.c
+++ b/src/FSAL/FSAL_GLUSTER/export.c
@@ -144,7 +144,13 @@ static fsal_status_t lookup_path(struct fsal_export *export_pub,
 		goto out;
 	}
 
-	glhandle = glfs_h_lookupat(glfs_export->gl_fs, NULL, realpath, &sb);
+#ifdef USE_GLUSTER_SYMLINK_MOUNT
+		glhandle = glfs_h_lookupat(glfs_export->gl_fs, NULL, realpath,
+					&sb, 1);
+#else
+		glhandle = glfs_h_lookupat(glfs_export->gl_fs, NULL, realpath,
+					&sb);
+#endif
 	if (glhandle == NULL) {
 		status = gluster2fsal_error(errno);
 		goto out;
diff --git a/src/FSAL/FSAL_GLUSTER/handle.c b/src/FSAL/FSAL_GLUSTER/handle.c
index 364a73d..7572b17 100644
--- a/src/FSAL/FSAL_GLUSTER/handle.c
+++ b/src/FSAL/FSAL_GLUSTER/handle.c
@@ -103,9 +103,13 @@ static fsal_status_t lookup(struct fsal_obj_handle *parent,
 	now(&s_time);
 #endif
 
-	glhandle =
-	    glfs_h_lookupat(glfs_export->gl_fs, parenthandle->glhandle, path,
-			    &sb);
+#ifdef USE_GLUSTER_SYMLINK_MOUNT
+		glhandle = glfs_h_lookupat(glfs_export->gl_fs,
+					parenthandle->glhandle, path, &sb, 0);
+#else
+		glhandle = glfs_h_lookupat(glfs_export->gl_fs,
+					parenthandle->glhandle, path, &sb);
+#endif
 	if (glhandle == NULL) {
 		status = gluster2fsal_error(errno);
 		goto out;
diff --git a/src/include/config-h.in.cmake b/src/include/config-h.in.cmake
index 9df3604..8e62531 100644
--- a/src/include/config-h.in.cmake
+++ b/src/include/config-h.in.cmake
@@ -51,6 +51,7 @@
 #cmakedefine HAVE_INCLUDE_LIBLUSTREAPI_H 1
 #cmakedefine HAVE_DAEMON 1
 #cmakedefine USE_LTTNG 1
+#cmakedefine USE_GLUSTER_SYMLINK_MOUNT 1
 
 #define NFS_GANESHA 1
 
-- 
1.7.1

